aliza: init at 1.9.7

luxzeitlos opened the pull request:
## Description of changes

We used to have aliza, but it was removed because it needed gtk4, was not build form source, and broke. We now build aliza from source and with gtk5.

We don't use by-name, because we need libsForQt5.callPackage.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

h7x4 commented:
> We don't use by-name, because we need libsForQt5.callPackage.

I think it still fits into `by-name`, if you `callPackage` the path of the package in `by-name`. There's a lot of new packages that do this to inherit from `darwin.apple_sdk.frameworks` for example.

h7x4 commented on pkgs/applications/science/medicine/aliza/default.nix:
```diff
@@ -0,0 +1,46 @@
+{ config, lib, pkgs, stdenv, wrapQtAppsHook, qtbase, ... }:
+
+stdenv.mkDerivation rec {
+  pname = "aliza";
+  version = "1.9.7";
+
+  src = pkgs.fetchFromGitHub {
+    owner = "AlizaMedicalImaging";
+    repo = "AlizaMS";
+    rev = "v${version}";
+    sha256 = "sha256-bambl3NP42KEwh6NPb8jhQlGqaUCSt6QuzUmmBnEAQw=";
+  };
+
+  nativeBuildInputs = [
+    wrapQtAppsHook
+  ];
+
+  buildInputs = [
+    pkgs.itk
+    pkgs.cmake
+    qtbase
+  ];
+
+  cmakeFlags = [
+    "-DCMAKE_BUILD_TYPE:STRING=Release"
+    "-DALIZA_QT_VERSION:STRING=5"
+    "-DITK_DIR:STRING=${pkgs.itk}"
```
I think it would be good to avoid importing `pkgs`

```suggestion
{ lib
, stdenv
, fetchFromGitHub
, wrapQtAppsHook
, itk
, cmake
, qtbase
}:

stdenv.mkDerivation rec {
  pname = "aliza";
  version = "1.9.7";

  src = fetchFromGitHub {
    owner = "AlizaMedicalImaging";
    repo = "AlizaMS";
    rev = "v${version}";
    sha256 = "sha256-bambl3NP42KEwh6NPb8jhQlGqaUCSt6QuzUmmBnEAQw=";
  };

  nativeBuildInputs = [
    wrapQtAppsHook
  ];

  buildInputs = [
    itk
    cmake
    qtbase
  ];

  cmakeFlags = [
    "-DCMAKE_BUILD_TYPE:STRING=Release"
    "-DALIZA_QT_VERSION:STRING=5"
    "-DITK_DIR:STRING=${itk}"
```

h7x4 commented on pkgs/applications/science/medicine/aliza/default.nix:
```diff
@@ -0,0 +1,46 @@
+    owner = "AlizaMedicalImaging";
+    repo = "AlizaMS";
+    rev = "v${version}";
+    sha256 = "sha256-bambl3NP42KEwh6NPb8jhQlGqaUCSt6QuzUmmBnEAQw=";
```
```suggestion
    hash = "sha256-bambl3NP42KEwh6NPb8jhQlGqaUCSt6QuzUmmBnEAQw=";
```

h7x4 commented on pkgs/applications/science/medicine/aliza/default.nix:
```diff
@@ -0,0 +1,46 @@
+
+  installPhase = ''
+    runHook preInstall
+    mkdir -p $out/bin/
+    cp ./bin/* $out/bin/
```
```suggestion
    install -Dm555 ./bin/* -t $out/bin
```

luxzeitlos committed:
aliza: init at 1.9.7

We used to have aliza, but it was removed because it needed gtk4, was not build form source, and broke. We now build aliza from source and with gtk5.

Co-authored-by: h7x4 <h7x4@nani.wtf>

luxzeitlos commented:
> > We don't use by-name, because we need libsForQt5.callPackage.
> 
> I think it still fits into `by-name`, if you `callPackage` the path of the package in `by-name`. There's a lot of new packages that do this to inherit from `darwin.apple_sdk.frameworks` for example.

@h7x4 thanks for the hint, but the validation does not like it if the second arg is empty.
Besides that this is stolen from `nextpnrWithGui`, which does the `libsForQt5.callPackage ../by-name...`.

Also this *does* work. Not sure if the validation is correct or not, or how to proceed here.

h7x4 commented:
> @h7x4 thanks for the hint, but the validation does not like it if the second arg is empty.

That seems like an oversight in the `pkgs/by-name` check.

@infinisil have you seen this before?

h7x4 commented:
Maybe it would make more sense to import `libsForQt5` as an argument in the package, and access the qt packages through `libsForQt5.<name>`?

infinisil commented:
The error could be improved, but it means that only packages for `pkgs.callPackage` are allowed, so no `libsForQt5.callPackage`. But indeed as @h7x4 suggests, you can likely refactor it to work with a standard `callPackage`, though I'm not 100% sure this would be exactly the same in the end.

> Besides that this is stolen from `nextpnrWithGui`, which does the `libsForQt5.callPackage ../by-name...`.

The extended definition of that is

```nix
  nextpnrWithGui = libsForQt5.callPackage ../by-name/ne/nextpnr/package.nix { ... }
```

Only attributes defined via `pkgs/by-name` need to use `callPackage`, there's no restrictions for other attributes. In this case, only `nextpnr` is defined in `pkgs/by-name`, `nextpnrWithGui` is not (it merely reuses the same file).

infinisil commented:
The error could be improved, but it means that only packages for `pkgs.callPackage` are allowed, so no `libsForQt5.callPackage`. But indeed as @h7x4 suggests, you can likely refactor it to work with a standard `callPackage`, though I'm not 100% sure this would be exactly the same in the end.

> Besides that this is stolen from `nextpnrWithGui`, which does the `libsForQt5.callPackage ../by-name...`.

The extended definition of that is

```nix
  nextpnrWithGui = libsForQt5.callPackage ../by-name/ne/nextpnr/package.nix { ... }
```

Only attributes defined via `pkgs/by-name` need to use `pkgs.callPackage`, there's no restrictions for other attributes. In this case, only `nextpnr` is defined in `pkgs/by-name` (via an implicit `pkgs.callPackage`), `nextpnrWithGui` is not (it merely reuses the same file).

luxzeitlos commented:
> Maybe it would make more sense to import `libsForQt5` as an argument in the package, and access the qt packages through `libsForQt5.<name>`?

How would that look like then?

luxzeitlos committed:
aliza: can't have it in by-name

luxzeitlos commented:
Well, so I took it out of by-name, should be good now?

AndersonTorres commented on pkgs/applications/science/medicine/aliza/default.nix:
```diff
@@ -0,0 +1,52 @@
+
+  buildInputs = [
+    itk
+    cmake
```
Usually it goes on nativeBuildInputs.

AndersonTorres commented on pkgs/applications/science/medicine/aliza/default.nix:
```diff
@@ -0,0 +1,52 @@
+    cmake
+    qtbase
+  ];
+
```
```suggestion

  strictDeps = true;

```

AndersonTorres commented on pkgs/applications/science/medicine/aliza/default.nix:
```diff
@@ -0,0 +1,52 @@
+  ];
+
+  cmakeFlags = [
+    "-DCMAKE_BUILD_TYPE:STRING=Release"
+    "-DALIZA_QT_VERSION:STRING=5"
+    "-DITK_DIR:STRING=${itk}"
+    "-DMDCM_USE_SYSTEM_ZLIB:BOOL=ON"
```
```suggestion
    (lib.cmakeFeature "CMAKE_BUILD_TYPE" "Release")
    (lib.cmakeFeature "ALIZA_QT_VERSION" "5")
    (lib.cmakeOptionType "filepath" "ITK_DIR" itk)
    (lib.cmakeBool "-MDCM_USE_SYSTEM_ZLIB" true)
```

luxzeitlos replied:
are you sure `(lib.cmakeBool "-MDCM_USE_SYSTEM_ZLIB" true)` is correct?

AndersonTorres replied:
... why not?

AndersonTorres commented on pkgs/applications/science/medicine/aliza/default.nix:
```diff
@@ -0,0 +1,52 @@
+  '';
+
+  meta = with lib; {
+    description = "Aliza MS DICOM Viewer";
```
```suggestion
    description = "A DICOM Viewer";
```

Do not be self-referent on description (except in cases of inside-jokes like  "GNU's Not Unix").

AndersonTorres commented on pkgs/applications/science/medicine/aliza/default.nix:
```diff
@@ -0,0 +1,52 @@
+, qtbase
+}:
+
+stdenv.mkDerivation rec {
```
Please consider `finalAttrs`:

```suggestion
stdenv.mkDerivation (finalAttrs: {
```

AndersonTorres commented on pkgs/applications/science/medicine/aliza/default.nix:
```diff
@@ -0,0 +1,52 @@
+  src = fetchFromGitHub {
+    owner = "AlizaMedicalImaging";
+    repo = "AlizaMS";
+    rev = "v${version}";
```
```suggestion
    rev = "v${finalAttrs.version}";
```

AndersonTorres commented:
> Maybe it would make more sense to import `libsForQt5` as an argument in the package, and access the qt packages through `libsForQt5.<name>`?

To be honest, I do not like that extra scoping.
I don't believe QT5/QT6 libraries are so important to the point of deserving an "overlay" like this.
X11 is way more important than this, and no one created a "libsForXorg.callPackage"  (yet).

Also it contaminates some decisions. I remember when cmake was wrapped under libsForQt5.callPackage...

luxzeitlos committed:
Update pkgs/applications/science/medicine/aliza/default.nix

Co-authored-by: Anderson Torres <torres.anderson.85@protonmail.com>

luxzeitlos committed:
Apply suggestions from code review

Co-authored-by: Anderson Torres <torres.anderson.85@protonmail.com>
